File: November2022/Java/SumOfSubarrayMinimums.java
// @saorav21994

class Solution {
    public int sumSubarrayMins(int[] arr) {
        Stack<Integer> stack = new Stack<>();
        stack.push(-1);     // index can not be < 0

        int res = 0;
        int MOD = 1_000_000_007;

        for (int i = 0; i <= arr.length; i++) {

            int cur = (i < arr.length) ? arr[i] : 0;    // minimum value of arr is 1, so 0 will make sure to process all the last element

            while (stack.peek() != -1 && cur < arr[stack.peek()]) {
                int stIdx = stack.pop();
                int leftBound = stIdx-stack.peek();
                int rightBound = i - stIdx;
                long value = (leftBound*rightBound*(long)arr[stIdx])%MOD;
                res += value;
                res %= MOD;
            }

            stack.push(i);

        }

        return res;
    }
}

File: November2022/C++ Sol/sum-of-subarray-minimums.cpp
class Solution {
public:
    int sumSubarrayMins(vector<int>& arr) {
    
        
        long long len = arr.size();
        long long ans = 0;
        long long mod = 1e9 + 7;
        
        vector<long long>left(len);
        vector<long long>right(len);
        stack<long long>st;
        
        for(int i=0;i<len;i++){
            
            while(!st.empty() && arr[st.top()]>arr[i]){
                st.pop();
            }
            
            
            if(st.empty()){
                 left[i] = i+1;
            }
            else{
            left[i] = i-st.top();
            }
            
            
            st.push(i);
        }
        
        while(!st.empty()){
            st.pop();
        }
        
         for(int i=0;i<len;i++){
             
            while(!st.empty() && arr[st.top()]>arr[i]){
                  right[st.top()] = i-st.top();
                st.pop();
            }
           
             
            st.push(i);
        }
        
        while(!st.empty()){
            
            right[st.top()] = len - st.top();
            st.pop();
        }
        
        for(long long i=0;i<len;i++){
         
            ans = ((ans%mod) + (left[i]*right[i]*arr[i]%mod))%mod;
        }
        
       
        return ans;
    }
};